### PR TITLE
feature(comments): river comment links go to right page and per page configurable

### LINKED ISSUE #7936: "view reply" river button
Shouldn't clicking view reply go to the appropriate pagination of the last commenter for discussion?

### PR DESCRIPTION
**Pull request by mrclay**
Fixes #7936 and #4193. Also makes comments per page configurable by hook. Although this is already configurable by the "view_vars" hook, the hook allows the comment URLs to point to the correct page.

### CODE REVIEW

**Comment by mrclay** [comment]
@hypeJunction curious how this would affect your Ajax work

**Comment by jdalsem** [comment]
LGTM

**Comment by mrclay** [comment]
FYI this adds a COUNT query for each comment in the river. A lot of our per work should help offset this.

**Comment by jdalsem** [comment]
out of the box thought:
why not have a link to `/comment/<guid>` that gets routed to the correct entity url (plus comments offset)?

**Comment by mrclay** [comment]
I was just typing that suggestion :)

**Comment by jdalsem** [comment]
offset could change over time, `comment/guid` will never change (unless the comment is removed, so maybe entity link should always be passed along)

That could also fix the "read the comment" link in the mail (that is crappy now, as it only links to the entity, and not the comment)

**Comment by mrclay** [comment]
Agreed. I guess we'd also want `/discussion/reply/GUID`.

**Comment by jdalsem** [comment]
related to #4193

**Comment by mrclay** [comment]
To be clear, $comment->getURL() would return `/comment/$comment_guid/$target_guid` and we should add another hook in the comment/ handler (or should we let plugin authors use the forward hook?).

File: engine/lib/comments.php [hunk 1]
```diff
@@ -85,6 +89,67 @@ function _elgg_comments_page_handler($page) {
...
+	$offset = floor($count / $limit) * $limit;
+	if (!$offset) {
+		$offset = null;
+	}
+
+	$url = elgg_http_add_url_query_elements($container->getURL(), [
+			'offset' => $offset,
+		]) . "#elgg-object-{$comment->guid}";
+
+	$url = _elgg_services()->hooks->trigger('forward_url', 'object/comment', null, $url);
```
Comment (mrclay) [inline engine/lib/comments.php]: Anyone know what to call this hook?

File: engine/lib/comments.php [hunk 2]
```diff
@@ -85,6 +89,67 @@ function _elgg_comments_page_handler($page) {
...
 /**
+ * Redirect to the comment in context of the containing page
+ *
+ * @param int $comment_guid  GUID of the comment
+ * @param int $fallback_guid GUID of the containing entity
+ *
+ * @return void
+ * @access private
+ */
+function _elgg_comment_redirect($comment_guid, $fallback_guid) {
```
Comment (mrclay) [inline engine/lib/comments.php]: These added functions are so similar but differ in so many ways I see no reason to put energy into refactoring.

File: engine/lib/comments.php (hunk 1 above)
Comment (juho-jaakkola) [inline engine/lib/comments.php]: I don't think we need this. We already have:
 - ```'entity:url', 'object'```
 - ```'forward', 'system'```

**Comment by mrclay** [comment]
I'd like to get one more core thumbs up on this idea.

**Comment by juho-jaakkola** [comment]
Tested and LGTM

**Comment by manacim** [comment]
+1

**Pull request merged**